### ISSUE TITLE
Sending JSON directly to ES with Filebeat

### ISSUE THREAD

**Issue by Sergej-Popov** [elastic/filebeat#311]
It would be nice if I could have my message written as object instead of string. I though I might be able to do it with mapping, but couldn't find the way.

Is this currently possible or I have to go via Logstash?

My log files always have correctly formatted JSON objects per line, so I want to bypass Logstash and write directly to ES since there is no enhancement required other then parsing string to JSON.
But whatever I do I always end up with message as string instead of object.

I have filebeat configured to read logs:
```
filebeat:
  prospectors:
      paths:
        - C:\logs\*\*.log
      input_type: log
     document_type: object
```

resulting event

```
{
      "_index" : "filebeat-2015.12.09",
      "_type" : "object",
      "_id" : "AVGHlgXkvPp1bej3HrM4",
      "_score" : 1.0,
      "_source":{"@timestamp":"2015-12-09T16:32:09.822Z","beat":{"hostname":"MACHINE-A","name":"MACHINE-A"},"count":1,"fields":null,"input_type":"log","message":"{ \"date\": \"2015-12-09 16:32:05,000 UTC\", \"message\": \"msg1\", \"level\": \"Info\"}","offset":933,"source":"C:\\logs\\test\\test.log","type":"object"}
    }
```

filebeat mapping:

```
{
  "filebeat" : {
    "order" : 0,
    "template" : "filebeat-*",
    "settings" : {
      "index" : {
        "refresh_interval" : "5s"
      }
    },
    "mappings" : {
      "_default_" : {
        "dynamic_templates" : [ {
          "template1" : {
            "mapping" : {
              "ignore_above" : 1024,
              "index" : "not_analyzed",
              "type" : "{dynamic_type}",
              "doc_values" : true
            },
            "match" : "*"
          }
        } ],
        "_all" : {
          "norms" : {
            "enabled" : false
          },
          "enabled" : true
        },
        "properties" : {
          "@timestamp" : {
            "type" : "date"
          },
          "offset" : {
            "type" : "long",
            "doc_values" : "true"
          },
          "message" : {
            "index" : "analyzed",
            "type" : "object"
          }
        }
      }
    },
    "aliases" : { }
  }
}
```

**Comment by urso** [comment]
parsing the json is not supported by filebeat. Any kind of parsing/enhancement must be done via Logstash.

There is some interesting progress on Elasticsearch maybe removing the need for LS in the future: https://github.com/elastic/elasticsearch/issues/14049

**Comment by Sergej-Popov** [comment]
@urso thanks for replying, but don't you feel like this a limitation for filebeat? I mean why it treats my log lines only as string? why can't I specify type for each of my log lines? say treat it as int or array or object. It makes sense that string is default, but nothing in line indicates it is a string, so it seem I should be able to specify. Maybe it is worth considering issue as future enhancement? My point is that it is not even a parsing feature, it is specifying type of message property. thanks.

**Comment by ruflin** [comment]
@Sergej-Popov As you mentioned correctly, it is possible with LS (but currently has a bug: https://github.com/logstash-plugins/logstash-input-beats/issues/33). We try to keep the feature set and complexity as minimal as possible. Even though it would probably possible to do most of the tasks of LS also on the client side, that is not the intention of the beats as also the resource usage should be kept low.

I think an interesting question here could be why the JSON object has to be written first to a file and then picked up? Is there a beat missing that could directly fetch the JSON object and forward it as such? Where is the JSON coming from?

In case there is quite a simple and stable implementation to send the actual JSON, it could be interesting. I worry we then have to deal with invalid JSON objects etc.

**Comment by Sergej-Popov** [comment]
@ruflin My scenario is simple: I just happen to output my log files in JSON format instead of arbitrary string. Why? Because JSON is NOT less readable, NOT harder to generate and this way my logs are ready for downstream processing. I actually think of this as best practice. But this is not the only scenario I can think of. Maybe I could be dumping performance counter data to csv file and therefore would want Filebeat to treat each line as array of numbers/strings.

Not having LS in between eliminates extra moving part to maintain. Don't get me wrong it is a nice tool but simple scenarios as above make it an overkill. But now a I have to use LS with JSON filter plugin.

I see your point about performance and validation. May be it could be solved simpler? Just as a first instinct: leave the validation to receiving end, Elastic Search validates JSON on write anyway? so filebeat could have a configuration option on how to deal with rejected message skip|retry|dropdead (I imagine there is some strategy for that already), in terms of performance it might be even faster, because you don't have to worry about escaping string if filebeat was configured to treat input as valid JSON object (of any type). Yes it might feel dangerous. But if this is how I decided to configure my system, then I have some level of confidence in my logs and it is not more error prone then parsing arbitrary string though regex patterns.

Thank you for consideration.

**Comment by ruflin** [comment]
I agree that if it is possible to have a structured JSON is superior and LS can be overkill sometimes. That is also why for example Topbeat and Packetbeat directly send structured data to elasticsearch. One idea could be to introduce `input_type:json` which borrows most functionality from `log` but has a json decoder on top. Like this the implementation would be independent. https://github.com/elastic/filebeat/blob/master/etc/filebeat.yml#L31

The problem is, that probably next is a request for csv, xml and others. 

What service is "logging" the service docs? Could there be an option that it not even has to be written to disk first but sent through a beat directly to elasticsearch? That would be my favorite option.

**Comment by kimchy** [comment]
I think it is relatively common now to see log output format of JSON to a file, ending in a new line. I agree with the concern of now needing to handle more outputs, but I am leaning towards thinking of maybe we should implement it, and only for JSON, as it is so common.

The tricky bit to me is how will the schema look like in this case..., what goes under `message`, the beat will also need to "enhance" the JSON with additional data, ... .

**Comment by ruflin** [comment]
@kimchy The easiest option would be to keep also the JSON document under message, means all the structure stays the same, but instead of a string under message with have a JSON document. So also no conflict fields can exist:

```
beat: { ...}
...
message: { JSON log content }
...
```
This would also be consistent with the way it is handled with the JSON codec in LS I think.

The open question would be what happens to the timestamp field. Would we overwrite the @timestamp if one exists in the json document?

**Comment by ruflin** [comment]
@Sergej-Popov I reopened the issue for further discussion.

**Comment by kimchy** [comment]
@ruflin I think we might want to look at how common logging libs output the data in JSON format (log4j, ruby, php, .NET), maybe if we can get some samples of how they look, we can make better suggestion for the schema... ?

**Comment by urso** [comment]
@ruflin no, it's not consistent with LS. In LS you want to use the json_line codec on your beats input.

The codec will read the 'message' field and parses the content as json. The parsed json is the top-level event to be index. All additional fields from original event will be written into parsed event (That is fields might be overwritten).

**Comment by ruflin** [comment]
@kimchy :+1: We should definitively further discuss the schema based on what is produced by others, especially which fields should be overwritten.
@urso Good to know.

**Comment by urso** [comment]
Still not sure I'd like to go this route. Having ingest nodes in ES will be much superior. In either case (json parsing in filebeat, LS, ES) the user is responsible for doing the mapping I would say. Maybe we can give some hints in doc.

**Comment by tsg** [comment]
Just to take a small step back here, wouldn't Node Ingest solve this in a more elegant way? I assume it will have the equivalent of a JSON codec. 

OTOH, having Filebeat understand the JSON has the advantage that it can filter data before sending it over the network.

Sounds like we'd need the equivalent of  the `fields_under_root` option we have for fields.

**Comment by kimchy** [comment]
I am not sure if node ingest would be able to easily handle this. There are no plans for "codecs" in node ingest, the assumption is that the requests comes in a valid bulk request format. We could have a parse JSON processor in node ingest to parse the json from `message` as a string and into another element or the same one, but then it means different pipelines for different inputs (JSON and non grok ones, ...).

I think the exercise of learning the typical schemas a JSON output produces will help learn more around how we want to approach this.

**Comment by Sergej-Popov** [comment]
From a usage perspective I like the idea of user being responsible for doing the mapping as I don't believe there is a standard output beyond Timestamp, Level and Message from logging libraries. But I guess not mapped fields should still be included, as there could be multiple message types. Consider this:
```
{"type": "LogEvents.ServiceStarted", "level": "Info"}
{"type": "LogEvents.ProcessingMessage", "level": "Debug"}
{"type": "LogEvents.Exceptions.MessageProcessingException", "level": "Warn", "reason": "Missing Id"}
```
So would I need to specify field `reason` in mapping? So I might later need to do the search for events of type MessageProcessingException where reason is XYZ, but not all events will have that field

**Comment by guyboertje** [comment]
@tsg, @ruflin, @urso, @kimchy - please be wary of upstream supplied JSON - especially if the JSON string is crafted (particularly from far end supplied strings) e.g. apache/nginx log format  and not generated by a JSON processor.  There is a strong risk of invalid JSON encoding of text. See these issues: [LS-4339](https://github.com/elastic/logstash/issues/4339#issuecomment-164325513) and [blog post comment](https://blog.pkhamre.com/logging-to-logstash-json-format-in-nginx/#comment-626580241)

Embarrassingly, we have an [example](https://github.com/elastic/examples/tree/master/ELK_nginx-json) that does not warn of the invalid JSON risk, i have created an issue though and will submit a PR.

**Comment by urso** [comment]
@guyboertje right. If we're to support JSON input it must be parsed and validated first!!! Beats must not forward invalid events/json!!!

**Comment by lookingcloudy** [comment]
Wow - this is the buggiest/worst software I have tried in a long time!  The assumptions are really bad.  The documentation is really bad.  The bugs are really bad.   Maybe pull this back internally until it is ready for public....or list it as alpha release.  Really frustrating.

**Comment by dedemorton** [comment]
@lookingcloudy  I'm sorry that your experience with the documentation was not positive. We have been making incremental improvements to the documentation over time, and additional improvements are underway. However, to help us address your specific concerns, it would be really useful to know exactly where you had problems. For example, did you find the basic setup instructions in the Getting Started topic hard to follow? Were there specific questions that weren't answered in the documentation? Were you looking for information that you could not find in the documentation? Were you confused by specific instructions? Which documentation was problematic (Filebeat, Logstash, etc)?

Please open a new issue (flagged with the docs label) to submit your feedback. Be as specific as possible so that we understand where you ran into trouble with the docs.

Thanks, and I look forward to seeing your feedback.

**Comment by urso** [comment]
Getting back on the original discussion.

@tsg @ruflin @kimchy @guyboertje One concern about supporting json logs directly from filebeat is, people assuming it will magically index all their json-logs (no matter what) in elasticsearch. Unfortunately with json-logs not necessarily having a proper schema and fields with same name having values of very different types (e.g. string vs. array of objects), we might fail to index more often than not. Correctly indexing json based logs might require quite some amount of processing and normalization, which is better done in logstash (still hell of a task, if developers don't stick to a common schema).

**Comment by ruflin** [comment]
@urso I think it is a very good point. Indexing JSON would only work properly if all JSON documents inside one log file have the same structure and each log file with a different structure is sent to a different type. This means, each json document type would need its own prospector.

In general I still think it would be interesting to crawl structured log data but we must find a good solution first.

**Comment by konrad-garus** [comment]
If it can be done with Logstash, Filebeat should support it as well. Yes, Elasticsearch does not like it when the same field has different types in logs. I have actually been bitten by this. It's perfectly fine though - my mistake, I fixed my logs, done.

It's better to have some support for structured data (should be obvious) and let developers fix their mistakes with regards to schema, than have no support. As it is, Filebeat is useless. If I still have to use Logstash for parsing, I can just as well use it for reading the files.

**Comment by mgcrea** [comment]
Big :+1: for native JSON support, should be a first-class citizen as it is so widely used. Also JSON being natively supported by go, so the complexity should not increase that much.

**Comment by RehanSaeed** [comment]
+1 Was making the switch to FileBeat but looks like it's just not ready.

**Comment by fatmanmclone90** [comment]
Please implement this

**Comment by guyboertje** [comment]
@mgcrea, @RehanSaeed, @fatmanmclone90 - How are you generating the JSON string that is put into your logs. Are you using a JSON generator in your application or the [nginx/apache log format hack?](https://blog.logentries.com/2014/08/json-logging-in-apache-and-nginx-with-logentries/) It is important that the JSON conforms exactly to the JSON spec - especially regarding escaping control characters.

[Read this comment](https://blog.logentries.com/2014/08/json-logging-in-apache-and-nginx-with-logentries/#comment-16369), person getting JSON parser errors because of badly escaped characters in the variable field values.

**Comment by RehanSaeed** [comment]
We are using the [Serilog](http://serilog.net/) logging framework to generate JSON logs. Each line in our logs looks something like this:

```
{"Timestamp":"2016-02-29T00:00:12.2323110+03:00","Level":"Debug","MessageTemplate":"Started GetAll","RenderedMessage":"Started GetAll","Properties":{"StartTime":"2016-02-28T21:00:12.2323110+00:00","SourceContext":"Some.Namesace.In.My.App","EnvironmentUserName":"DOMAIN\\Some.User","MachineName":"MACHINE-NAME","ProcessId":9984,"ThreadId":8,"Source":"My Application"}}
```

Most of our logs have the same JSON properties as the above line. When an exception occurs or when we provide additional custom information in our logs, we do have additional properties that appear under the `Properties' section.

**Comment by mgcrea** [comment]
For us, we are using the NodeJS package [Bunyan](https://github.com/trentm/node-bunyan), our backend directly writes logs (NginX is only used as an HTTPS proxy).

From the README:

**Bunyan log records are JSON.** A few fields are added automatically:
"pid", "hostname", "time" and "v".

```sh
$ node hi.js
{"name":"myapp","hostname":"banana.local","pid":40161,"level":30,"msg":"hi","time":"2013-01-04T18:46:23.851Z","v":0}
{"name":"myapp","hostname":"banana.local","pid":40161,"level":40,"lang":"fr","msg":"au revoir","time":"2013-01-04T18:46:23.853Z","v":0}
```

**Comment by guyboertje** [comment]
@mgcrea, @RehanSaeed - this is good information for us to know. Thanks.

**Comment by vnazarenko** [comment]
+1 for json processing on filebeat side.

**Comment by mfriedenhagen** [comment]
In the Java world there are a some of "enhancers" for logging which will produce JSON:
* https://github.com/logstash/logstash-logback-encoder
* https://github.com/SYNAXON/logstash-util-formatter
You might configure the Java applications to do logging directly, but this introduces another direct dependency to logstash. Writing JSON intermediately to a file or (e.g. with containers) to stderr decouples stuff IMO.

**Comment by ruflin** [comment]
For reference: @tsg Has done two implementation of JSON here:
* https://github.com/elastic/beats/pull/1069
* https://github.com/elastic/beats/pull/1143

**Comment by fatmanmclone90** [comment]
Has this functionality been released in Filebeats 1.1.2?  If so, is there any guidance on how to setup the config and templates?

**Comment by tsg** [comment]
@fatmanmclone90 No, it wasn't. https://github.com/elastic/beats/pull/1143 was only merged in master last week. We're going to have a 5.0-alpha1, most likely next week, and it will be in that. But, well, alpha.

**Comment by kylecordes** [comment]
It seems like a number of us have a similar use case. We are already emitting lines of log which consist of JSON in the exact form it needs to land in elastic. We have a logstash running somewhere, using a surprising amount of RAM, performing basically no processing as it takes these lines of JSON and shove them into elastic. Even an alpha version of something that eliminates that large complex system in the middle, should be adequate and perhaps more robust. Logstash is great for doing its job, but it basically has no job to do in this use case except help the global RAM market...

**Comment by ruflin** [comment]
Closing this issue as this feature was now released as part of the 5.0.0 releases

**Comment by fatmanmclone90** [comment]
Great news that it's released, is it documented anywhere how to configure it?

**Comment by ruflin** [comment]
https://www.elastic.co/guide/en/beats/filebeat/master/configuration-filebeat-options.html#config-json

**Comment by tsg** [comment]
@fatmanmclone90 To clarify, it's released in 5.0.0-alpha. There is still a bit of time until it makes it in a GA release.

**Comment by vortex314** [comment]
@elastic.co 
We are also looking for a simple solution to get logs into ES.
Our chain looks like this ( based on standardized JSON log format ) :+1: 
Application -> JSON log file -> FileBeat -> Kafka -> LogStash ( no parsing ) -> ES
or 
Application via Log4J KafkaAppender -> Kafka -> LS -> ES
Just to avoid this complex parsing depending on the application. Using the same data format everywhere strongly improves correlations.

**Comment by ruflin** [comment]
@vortex314 As this issue is closed, can you please open a new topic here for discussions / questions: https://discuss.elastic.co/c/beats/filebeat ?
